File: Automation Testing Python/Sample Programs/Factorial.py
# 4 = 4 * 3 * 2 * 1 
# 5 = 5 * 4 * 3 * 2 * 1

# print("Factorial of 5 : ", 5 * 4 * 3 * 2 * 1)

# ## -----------------------------------------------------
# # Function start
# def factorial(): # Fun def 
#     print("factorial code")
#     #other statements
#     #other code
# #Function end

# factorial()     # function call 


# ## -----------------------------------------------------
# # Function start
# def factorial(): # Fun def 
#     print("factorial code")
#     #other statements
#     #other code
#     factorial()
# #Function end

# factorial()     # function call 

# ## --------------------------------------------

# Factorial program with recursion 
# 5 = 5 * 4 * 3 * 2 * 1

print("Factorial of 8 : ", 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1)

def factorial(n : int):
    if n <= 1:     # n == 1 or n == 0
        return 1
    else: 
        return n * factorial(n-1)       #n * (n-1) * (n-2) * (n-3) * (n-4)

# 3 * 2 * 1

## -- function call 
num = -2
fact = factorial(num)
print(f"Factorial of {num} : {fact}")

# 3 * 2 * 1

File: JS_TS-main/JS/programmes/factorial.js

function factorial(n) {
    if(n === 1) {
        return 1;
    }

    return n * factorial(n - 1);
}

console.log(factorial(5));



//Factorial program:
//Approach 1:
/*
    5! = (5-0) * (5-1)  * (5-2) 60 * (5-3) * (5-4);
*/
function getFactorialOfNum(num){
    let c = num;
    for(let i=1; i<num; i++){
        c = c * (num - i);
    }
    console.log(c);
}
getFactorialOfNum(4);


//Approach 2 --> Best approach:
function fact(num){
    if(num<0){
        return `The ${num} is a -ve number. Please pass a +ve number.`;
    }
    if(num == 0 || num == 1){
        return 1;
    }
    return num * fact(num - 1);
}
console.log(fact(-1));
